src/mongocrypt-marking.c

```diff
@@ -1164,72 +1162,130 @@ static bool _fle2_generate_TextSearchTokenSets(_mongocrypt_key_broker_t *kb,
+                    _mongocrypt_buffer_cleanup(&asBsonValue);
+                    mc_TextSuffixTokenSet_cleanup(&tset);
+                    goto fail;
+                }
+                _mc_array_append_val(&tsts->suffixArray, tset); // moves ownership of tset
+                _mongocrypt_buffer_cleanup(&asBsonValue);
+            }
         }
-        _mc_array_append_val(&tsts->suffixArray, suffixSet);
     }
+
+    // Generate array of prefix token sets
     if (spec->prefix.set) {
-        // TODO MONGOCRYPT-762 iterate on StrEncode suffixes set
-        mc_TextPrefixTokenSet_t prefixSet = {{0}};
-        mc_TextPrefixTokenSet_init(&prefixSet);
-
-        if (!_fle2_generate_TextPrefixTokenSet(kb,
-                                               &prefixSet,
-                                               &foldedValue,
-                                               contentionFactor,
-                                               common.collectionsLevel1Token,
-                                               common.serverTokenDerivationLevel1Token,
-                                               status)) {
-            mc_TextPrefixTokenSet_cleanup(&prefixSet);
-            goto fail;
+        mc_affix_set_iter_t set_itr;
+        mc_affix_set_iter_init(&set_itr, encodeSets->prefix_set);
+
+        while (mc_affix_set_iter_next(&set_itr, &substring, &bytelen, &appendCount)) {
+            for (; appendCount > 0; appendCount--) {
+                mc_TextPrefixTokenSet_t tset = {{0}};
+                mc_TextPrefixTokenSet_init(&tset);
+
+                _mongocrypt_buffer_t asBsonValue;
+                _mongocrypt_buffer_init(&asBsonValue);
+                BSON_ASSERT(bytelen < INT_MAX);
+                _mongocrypt_buffer_copy_from_string_as_bson_value(&asBsonValue, substring, (int)bytelen);
+
+                if (!_fle2_generate_TextPrefixTokenSet(kb,
+                                                       &tset,
+                                                       &asBsonValue,
+                                                       contentionFactor,
+                                                       common.collectionsLevel1Token,
+                                                       common.serverTokenDerivationLevel1Token,
+                                                       status)) {
+                    _mongocrypt_buffer_cleanup(&asBsonValue);
+                    mc_TextPrefixTokenSet_cleanup(&tset);
+                    goto fail;
+                }
+                _mc_array_append_val(&tsts->prefixArray, tset); // moves ownership of tset
+                _mongocrypt_buffer_cleanup(&asBsonValue);
+            }
         }
-        _mc_array_append_val(&tsts->prefixArray, prefixSet);
     }
     payload->textSearchTokenSets.set = true;
     res = true;
 fail:
     _FLE2EncryptedPayloadCommon_cleanup(&common);
-    _mongocrypt_buffer_cleanup(&foldedValue);
+    mc_str_encode_sets_destroy(encodeSets);
```

Comment:
Don't we have to also clean up tsts, since it's holding a bunch of initialized buffers?

Reply:
they get cleaned up when the caller of this calls `mc_FLE2InsertUpdatePayloadV2_cleanup`.